Space varying material properties

dalg24 opened the pull request.

Commit:
Pass fe_values instead of cell in MPValues::get_values()

Commit:
Add FunctionSpaceMPValues

Commit:
Add to_map utility to parse input

Commit:
Add new constructor to FunctionSpaceMPValues

Commit:
Add test for space varying material properties

Commit:
Bugfix (parse empty maps)

Rombur commented on cpp/source/deal.II/mp_values.h:
```diff
@@ -101,6 +93,26 @@ class UniformConstantMPValues : public MPValues<dim>
+                  std::vector<double> &values) const override;
+
+private:
+  std::shared_ptr<dealii::Function<dim>> const _function;
```
Why do you use a Function and not FunctionParser?

dalg24 replied:
more general.  we could consider move uniform constant to simple `dealii::ConstantFunction<dim>`

Rombur commented on cpp/source/utils.templates.h:
```diff
@@ -8,12 +8,25 @@
 namespace cap
 {
 
+namespace impl
```
What not `namespace internal`? Do we need to have a different namespace?

dalg24 replied:
k will change it

Rombur commented on cpp/source/utils.templates.h:
```diff
@@ -47,4 +54,28 @@ std::string to_string(std::vector<T> const &v)
+    if (kv.size() != 2)
+      throw std::runtime_error("invalid key-value pair " + p);
+    for (auto &x : kv)
+      boost::algorithm::trim(x);
```
Does this get ride of blank space?

dalg24 replied:
strip whitespace left and right

Rombur commented on cpp/test/test_mp_values.cc:
```diff
@@ -14,9 +14,85 @@
+  std::vector<dealii::Point<dim>> points = {{0, 0}, {0, 1}, {0.5, 0.5}, {1, 1}};
+  // NOTE: I tried with dealii::update_default but I got a memory access
+  // violation instead of some deal.II exception as I expected...
+  // I was using deal.II in release mode.
```
There are very few checks in release mode. That's why it goes at least five times faster.

Commit:
Add test for custom liquid electrical conductivity

dalg24 commented:
@Rombur I changed the namespace and committed the switch that allows us to overwrite \kappa in the electrode and the separator.  All we need is adding a node `custom_liquid_electrical_conductivity { expression: ..., variables[optional]: ..., constants[optional]: ... }` in `material_properties.{electrode,separator}_material`.

Rombur commented:
tests don't pass

Commit:
bugfix in equivalent circuit

Commit:
Temporary bugfix in Postprocessor

The issue is that MPValues cannot evaluate values on FEFaceValues at this point.
Also made fe_values and fe_face_values update the quadrature points by default.

Commit:
Updated test_mp_values.cpp

Since it is not valid any more to pass empty values to MPValues.
It used to be fine since the get_values() was only a std::fill() for
UniformConstantMPValues and there was no check on the size.

dalg24 commented:
ping.  in particular, look at c7a4051.

Commit:
Bugfix session in Docker container

dalg24 commented:
only the "timer" test failed so I'd say this is all good

Rombur commented on cpp/source/deal.II/mp_values.h:
```diff
@@ -101,6 +93,26 @@ class UniformConstantMPValues : public MPValues<dim>
+class FunctionSpaceMPValues : public MPValues<dim>
+{
+public:
+  FunctionSpaceMPValues(std::shared_ptr<dealii::Function<dim>> const &function);
```
`const &`? Also add blank line between the two constructors.

dalg24 replied:
Yeah, the `shared_ptr` is still an object;)  You want me to pass by value?

Rombur commented on cpp/source/utils.templates.h:
```diff
@@ -8,12 +8,25 @@
+  if (b.compare("true") == 0)
+    b = "1";
+  else if (b.compare("false") == 0)
+    b = "0";
```
This should throw an exception in the `else` case.

dalg24 replied:
Not necessary, the `boost::lexical_cast<boost>()` will throw otherwhise

Rombur reviewed (approved):
Looks good please fix the two small comments and merge yourself.

Commit:
Fix const correctness in FunctionSpaceMPValues
